crates/bevy_ecs/src/relationship/relationship_query.rs

```diff
@@ -270,3 +276,356 @@ where
+                &state.filter_state,
+                last_run,
+                this_run,
+            ),
+        }
+    }
+
+    const IS_DENSE: bool = <F as WorldQuery>::IS_DENSE & <&R as WorldQuery>::IS_DENSE;
+
+    unsafe fn set_archetype<'w>(
+        fetch: &mut Self::Fetch<'w>,
+        state: &Self::State,
+        archetype: &'w Archetype,
+        table: &'w Table,
+    ) {
+        <&'static R as WorldQuery>::set_archetype(
+            &mut fetch.relation_fetch,
+            &state.relation_state,
+            archetype,
+            table,
+        );
+        <F as WorldQuery>::set_archetype(
+            &mut fetch.filter_fetch,
+            &state.filter_state,
+            archetype,
+            table,
+        );
+    }
+
+    unsafe fn set_table<'w>(fetch: &mut Self::Fetch<'w>, state: &Self::State, table: &'w Table) {
+        <&'static R as WorldQuery>::set_table(
+            &mut fetch.relation_fetch,
+            &state.relation_state,
+            table,
+        );
+        <&R as WorldQuery>::set_table(&mut fetch.relation_fetch, &state.relation_state, table);
+        <F as WorldQuery>::set_table(&mut fetch.filter_fetch, &state.filter_state, table);
+    }
+
+    unsafe fn fetch<'w>(
+        fetch: &mut Self::Fetch<'w>,
+        entity: Entity,
+        table_row: TableRow,
+    ) -> Self::Item<'w> {
+        // Look up the relationship
+        // SAFETY: the safety requirements for calling `fetch` on `R` are a subset of the safety requirements for calling this method
+        let relation = unsafe {
+            <&'static R as WorldQuery>::fetch(&mut fetch.relation_fetch, entity, table_row)
+        };
+        // Then figure out what the related entity is
+        let related_entity = relation.get();
+
+        // Finally, check if the related entity matches the filter
+        // SAFETY: the safety requirements for calling `fetch` on `F` are a subset of the safety requirements for calling this method
+        unsafe { <F as WorldQuery>::fetch(&mut fetch.filter_fetch, related_entity, table_row) }
+    }
+
+    fn update_component_access(
+        state: &Self::State,
+        access: &mut crate::query::FilteredAccess<crate::component::ComponentId>,
+    ) {
+        <&'static R as WorldQuery>::update_component_access(&state.relation_state, access);
+        <F as WorldQuery>::update_component_access(&state.filter_state, access);
+    }
+
+    fn init_state(world: &mut crate::prelude::World) -> Self::State {
+        RelatedFilterState {
+            relation_state: <&'static R as WorldQuery>::init_state(world),
+            filter_state: <F as WorldQuery>::init_state(world),
+        }
+    }
+
+    fn get_state(components: &crate::component::Components) -> Option<Self::State> {
+        Some(RelatedFilterState {
+            relation_state: <&'static R as WorldQuery>::get_state(components)?,
+            filter_state: <F as WorldQuery>::get_state(components)?,
+        })
+    }
+
+    fn matches_component_set(
```

Comment:
Given the test pattern, I suspect that this method is wrong, but I can't figure out why.